server/src/main/java/org/elasticsearch/cluster/routing/allocation/command/MoveAllocationCommand.java

```diff
@@ -102,7 +102,16 @@ public RerouteExplanation execute(RoutingAllocation allocation, boolean explain)
         Decision decision = null;
 
         boolean found = false;
-        for (ShardRouting shardRouting : allocation.routingNodes().node(fromDiscoNode.getId())) {
+        RoutingNode fromRoutingNode = allocation.routingNodes().node(fromDiscoNode.getId());
+        if (fromRoutingNode == null && !fromDiscoNode.isDataNode()) {
+            throw new IllegalArgumentException("[move_allocation] can't move " + shardId + ", from " + fromDiscoNode + ", to " + toDiscoNode + ", since its not allowed, reason: " +  fromDiscoNode + " is not a data node.");
```

Comment:
Nit: I think it'd be better for the message to read:

```[move_allocation] can't move abc123 from node1 to node2: node1 is not a data node```

(NB less punctuation, and no need to say `since its not allowed`)